File: android/src/main/kotlin/io/wisetrack/wisetrack/WisetrackPlugin.kt
package io.wisetrack.wisetrack

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.wisetrack.sdk.core.WiseTrack
import io.wisetrack.sdk.core.models.DeeplinkCallback
import io.wisetrack.sdk.core.models.EventParam
import io.wisetrack.sdk.core.models.RevenueCurrency
import io.wisetrack.sdk.core.models.WTEvent
import io.wisetrack.sdk.core.models.WTEventType
import io.wisetrack.sdk.core.models.WTInitialConfig
import io.wisetrack.sdk.core.models.WTLogLevel
import io.wisetrack.sdk.core.models.WTStoreName
import io.wisetrack.sdk.core.models.WTUserEnvironment
import io.wisetrack.sdk.core.utils.wrapper.ResourceWrapper


/** WiseTrackPlugin */
class WiseTrackPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "io.wisetrack.flutter")
        channel.setMethodCallHandler(this)

        WiseTrack.ensureInitialized(context)
        WiseTrack.addLoggerOutput(FlutterChannelOutputLogger(channel))

//        application = flutterPluginBinding.applicationContext as Application
//        application?.registerActivityLifecycleCallbacks(lifecycleHandler)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        WiseTrack.ensureInitialized(context)

        when (call.method) {
            MethodNames.INIT -> {
                initSDK(call)
                result.success(null)
            }

            MethodNames.CLEAR_AND_STOP -> {
                clearDataAndStop()
                result.success(null)
            }

            MethodNames.SET_LOG_LEVEL -> {
                val level = call.argument<Int>("level") ?: 0
                setLogLevel(level)
                result.success(null)
            }

            MethodNames.SET_ENABLED -> {
                val enabled = call.argument<Boolean>("enabled") ?: true
                setEnabled(enabled)
                result.success(null)
            }

            MethodNames.IS_ENABLED -> {
                val enabled = isEnabled()
                result.success(enabled)
            }

            MethodNames.IOS_REQUEST_FOR_ATT -> {
                // Not for android platform
                result.success(false)
            }

            MethodNames.START_TRACKING -> {
                startTracking()
                result.success(null)
            }

            MethodNames.STOP_TRACKING -> {
                stopTracking()
                result.success(null)
            }

            MethodNames.SET_APNS_TOKEN -> {
                // APNS is not used in Android!
                result.success(null)
            }

            MethodNames.SET_FCM_TOKEN -> {
                val token = call.argument<String>("token")
                setFCMToken(token)
                result.success(null)
            }

            MethodNames.SET_PACKAGES_INFO -> {
                setPackagesInfo()
                result.success(null)
            }

            MethodNames.LOG_EVENT -> {
                logEvent(call)
                result.success(null)
            }

            MethodNames.GET_ADID -> {
                // Not for android platform
                val adid = getAdId()
                result.success(adid)
            }

            MethodNames.GET_IDFA -> {
                // Not for android platform
                result.success(null)
            }

            MethodNames.GET_REFERRER -> {
                val referrer = getReferrer()
                result.success(referrer)
            }

            MethodNames.GET_LAST_DEEPLINK -> {
                val link = getLastDeeplink()
                result.success(link)
            }

            MethodNames.GET_DEFERRED_LINK -> {
                val link = getDeferredDeeplink()
                result.success(link)
            }

            MethodNames.IS_WISETRACK_NOTIFICATION -> {
                val isWiseTrackNotification = isWiseTrackNotification(call)
                result.success(isWiseTrackNotification)
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    private fun initSDK(call: MethodCall) {

        val resourceWrapper = ResourceWrapper(context)
        resourceWrapper.setFramework("flutter")
        resourceWrapper.setEnvironment(call.argument<String>("sdk_env")!!)
        resourceWrapper.setVersion(call.argument<String>("sdk_version")!!)

        val initialConfig = WTInitialConfig(
            appToken = call.argument<String>("app_token")!!,
            clientSecret = call.argument<String>("client_secret")!!,
            environment = WTUserEnvironment.valueOf(
                call.argument<String>("user_environment")!!.uppercase()
            ),
            storeName = WTStoreName.fromString(call.argument<String>("android_store_name")!!),
            trackingWaitingTime = call.argument<Int>("tracking_waiting_time")!!,
            startTrackerAutomatically = call.argument<Boolean>("start_tracker_automatically")!!,
            customDeviceId = call.argument<String?>("custom_device_id"),
            defaultTracker = call.argument<String?>("default_tracker"),
            deeplinkEnabled = call.argument<Boolean?>("deeplink_enabled") ?: true,
            logLevel = WTLogLevel.fromPriority(call.argument<Int>("log_level")!!),
            oaidEnabled = call.argument<Boolean>("oaid_enabled") ?: false,
        )

        WiseTrack.initialize(context, initialConfig)

        // Set deeplink listener after initialization
        WiseTrack.setOnDeeplinkListener { intent, isDeferred ->
            Handler(Looper.getMainLooper()).post {
                channel.invokeMethod(
                    MethodNames.DEEPLINK_LISTENER,
                    mapOf(
                        "url" to intent.dataString,
                        "is_deferred" to isDeferred
                    ),
                )
            }
        }
    }

    private fun clearDataAndStop() {
        /// Clear all caches and data
        WiseTrack.clearDataAndStop()
    }

    private fun setLogLevel(level: Int) {
        WiseTrack.setLogLevel(WTLogLevel.fromPriority(level))
    }

    private fun setEnabled(enabled: Boolean) {
        WiseTrack.setEnabled(enabled)
    }

    private fun isEnabled(): Boolean {
        return WiseTrack.isEnabled
    }

    private fun startTracking() {
        WiseTrack.startTracking()
    }

    private fun stopTracking() {
        WiseTrack.stopTracking()
    }

    private fun setFCMToken(token: String?) {
        if (token != null)
            WiseTrack.setFCMToken(token)
    }

    private fun setPackagesInfo() {
        WiseTrack.setPackagesInfo()
    }

    private fun logEvent(call: MethodCall) {
        val eventType = WTEventType.valueOf(call.argument<String>("type")!!.uppercase())
        val eventName = call.argument<String>("name")!!
        val eventParam = call.argument<Map<String, Any>>("params")?.mapValues {
            when (it.value) {
                is Int -> EventParam((it.value as Int).toDouble())
                is Double -> EventParam(it.value as Double)
                is Boolean -> EventParam(it.value as Boolean)
                else -> EventParam(it.value.toString())
            }
        }

        val event = when (eventType) {
            WTEventType.DEFAULT -> {
                WTEvent.defaultEvent(eventName, params = eventParam)
            }

            WTEventType.REVENUE -> {
                WTEvent.revenueEvent(
                    eventName,
                    amount = call.argument<Double>("revenue")!!,
                    currency = RevenueCurrency.valueOf(
                        call.argument<String>("currency")!!.uppercase()
                    ),
                    params = eventParam
                )
            }
        }
        WiseTrack.logEvent(event)
    }

    private fun getAdId(): String? {
        return WiseTrack.getADID()
    }

    private fun getReferrer(): String? {
        return WiseTrack.getReferrer()
    }

    private fun getLastDeeplink(): String? {
        return WiseTrack.getLastDeeplink()
    }

    private fun getDeferredDeeplink(): String? {
        return WiseTrack.getDeferredDeeplink()
    }

    private fun isWiseTrackNotification(call: MethodCall): Boolean? {
        try {
            if (call.arguments !is Map<*, *>) return false
            val payload = (call.arguments as Map<*, *>)
                .mapKeys { (key, _) -> key.toString() }
                .mapValues { (_, value) -> value.toString() }
            return WiseTrack.isWiseTrackNotificationPayload(payload)

        } catch (_: Exception) {
        }
        return false
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/WisetrackPlugin.swift
import Flutter
import UIKit
import WiseTrackLib

public class WisetrackPlugin: NSObject, FlutterPlugin {
    private var channel: FlutterMethodChannel?
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = WisetrackPlugin()
        let channel = FlutterMethodChannel(name: "io.wisetrack.flutter", binaryMessenger: registrar.messenger())
        instance.channel = channel
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.addApplicationDelegate(instance)
        
        WiseTrack.shared.prepareInitialization()
        WTLogger.shared.addOutput(output: FlutterChannelOutputLogger(channel: channel))
    }
    
    private func args(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> [String: Any]?{
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments for \(call.method)", details: nil))
            return nil
        }
        return args
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        
        switch call.method {
        case WisetrackMethodChannel.initSDK.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.initSDK(args: args)
            result(nil)
            
        case WisetrackMethodChannel.clearAndStop.rawValue:
            self.clearAndStop()
            result(nil)
            
        case WisetrackMethodChannel.setLogLevel.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.setLogLevel(level: args["level"] as! Int)
            result(nil)
            
        case WisetrackMethodChannel.setEnabled.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.setEnabled(enabled: args["enabled"] as! Bool)
            result(nil)
            
        case WisetrackMethodChannel.isEnabled.rawValue:
            let isEnable = self.isEnabled()
            result(isEnable)
            
        case WisetrackMethodChannel.iOSRequestForATT.rawValue:
            guard let _ = Bundle.main.object(forInfoDictionaryKey: "NSUserTrackingUsageDescription") as? String else {
                result(FlutterError(code: "MISSING_INFO_PLIST_KEY",
                                    message: "NSUserTrackingUsageDescription key is missing in Info.plist",
                                    details: "Please add NSUserTrackingUsageDescription to your Info.plist"))
                return
            }
            
            self.iOSRequestForATT(completion: { isAuthorized in
                result(isAuthorized)
            })
            
        case WisetrackMethodChannel.getADID.rawValue:
            result(nil)
            
        case WisetrackMethodChannel.getIDFA.rawValue:
            let res = self.getIDFA()
            result(res)
            
        case WisetrackMethodChannel.startTracking.rawValue:
            self.startTracking()
            result(nil)
            
        case WisetrackMethodChannel.stopTracking.rawValue:
            self.stopTracking()
            result(nil)
            
        case WisetrackMethodChannel.setAPNSToken.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.setAPNSToken(apnsToken: args["token"] as! String)
            result(nil)
            
        case WisetrackMethodChannel.setFCMToken.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.setFCMToken(fcmToken: args["token"] as! String)
            result(nil)
            
        case WisetrackMethodChannel.logEvent.rawValue:
            guard let args = args(call, result: result) else { return }
            
            self.logEvent(args: args)
            result(nil)
            
        case WisetrackMethodChannel.isWiseTrackNotification.rawValue:
            guard let args = args(call, result: result) else { return }
            result(self.isWiseTrackNotification(args: args))
            
        case WisetrackMethodChannel.getLastDeeplink.rawValue:
            let res = self.getLastDeeplink()
            result(res)
            
        case WisetrackMethodChannel.getDeferredLink.rawValue:
            let res = self.getDeferredLink()
            result(res)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension WisetrackPlugin {
    
    private func initSDK(args: [String: Any]) {
        ResourceWrapper.setSdkEnvironment(env: args["sdk_env"] as! String)
        ResourceWrapper.setSdkFramework(framework: "flutter")
        ResourceWrapper.setSdkVersion(version: args["sdk_version"] as! String)
        
        WiseTrack.shared.initialize(with: WTInitialConfig(
            appToken: args["app_token"] as! String,
            clientSecret: args["client_secret"] as! String,
            storeName: WTStoreName(rawValue: args["ios_store_name"] as! String),
            environment: WTUserEnvironment(rawValue: args["user_environment"] as! String) ?? .production,
            logLevel: WTLogLevel(rawValue: args["log_level"] as! Int),
            trackingWaitingTime: TimeInterval(args["tracking_waiting_time"] as! Int),
            startTrackerAutomatically: args["start_tracker_automatically"] as! Bool,
            customDeviceId: args["custom_device_id"] as? String,
            defaultTracker: args["default_tracker"] as? String,
            deeplinkEnabled: args["deeplink_enabled"] as! Bool,
            attWaitingInterval: args["att_waiting_interval"] as? TimeInterval,
            requestATTAutomatically: args["request_att_automatically"] as! Bool,
        ))

        // Set deeplink listener after initialization
        WiseTrack.shared.onDeeplinkReceived { url, isDeferred in
            DispatchQueue.main.async {
                self.channel?.invokeMethod(WisetrackMethodChannel.deeplinkListener.rawValue, arguments: [
                    "url": url.absoluteString, "is_deferred": isDeferred
                ])
            }
        }
    }
    
    private func isWiseTrackNotification(args: [String: Any]) -> Bool{
        return WiseTrack.shared.isWiseTrackNotificationPayload(userInfo: args)
    }
    
    private func clearAndStop() {
        WiseTrack.shared.clearDataAndStop()
    }
    
    private func setLogLevel(level: Int) {
        WiseTrack.shared.setLogLevel(WTLogLevel(rawValue: level) ?? WTLogLevel.warning)
    }
    
    private func setEnabled(enabled: Bool) {
        WiseTrack.shared.setEnabled(enabled: enabled)
    }
    
    private func isEnabled() -> Bool {
        return WiseTrack.shared.isEnabled
    }
    
    private func iOSRequestForATT(completion: @escaping (_ isAuthorized: Bool) -> Void) {
        WiseTrack.shared.requestAppTrackingAuthorization(completion: completion)
    }
    
    private func startTracking() {
        WiseTrack.shared.startTracking()
    }
    
    private func stopTracking() {
        WiseTrack.shared.stopTracking()
    }
    
    private func setAPNSToken(apnsToken: String) {
        WiseTrack.shared.setAPNSToken(token: apnsToken)
    }
    
    private func setFCMToken(fcmToken: String) {
        WiseTrack.shared.setFCMToken(token: fcmToken)
    }
    
    private func getIDFA() -> String? {
        return WiseTrack.shared.getIDFA()
    }
    
    private func logEvent(args: [String: Any]) {
        let params: [String: JSONValue]? = (args["params"] as? [String: Any])?.mapValues({ JSONValue.convert(value: $0)})
        
        switch args["type"] as! String {
        case WTEventType.default.name:
            WiseTrack.shared.logEvent(WTEvent.default(for: args["name"] as! String,
                                                      params: params))
        case WTEventType.revenue(currency: RevenueCurrency.USD, amount: 0).name:
            WiseTrack.shared.logEvent(WTEvent.revenue(for: args["name"] as! String,
                                                      currency: RevenueCurrency(rawValue: args["currency"] as! String)!,
                                                      amount: args["revenue"] as! Double,
                                                      params: params))
        default:
            return
        }
    }
    
    private func getLastDeeplink() -> String? {
        return WiseTrack.shared.getLastDeeplink()
    }
    
    private func getDeferredLink() -> String? {
        return WiseTrack.shared.getDeferredDeeplink()
    }
}

// MARK: - Deep Link Handling
extension WisetrackPlugin {

    // Handle Universal Links (AppDelegate)
    public func application(
        _ application: UIApplication,
        continue userActivity: NSUserActivity,
        restorationHandler: @escaping ([Any]) -> Void
    ) -> Bool {
        guard userActivity.activityType == NSUserActivityTypeBrowsingWeb,
              let url = userActivity.webpageURL else {
            return false
        }
        WiseTrack.shared.handleDeepLink(with: url)
        return true
    }

    // Handle Custom URL Schemes (AppDelegate)
    public func application(
        _ application: UIApplication,
        open url: URL,
        options: [UIApplication.OpenURLOptionsKey : Any] = [:]
    ) -> Bool {
        WiseTrack.shared.handleDeepLink(with: url)
        return true
    }
}

extension Encodable {
    
    /// Converting object to postable dictionary
    func toDictionary(_ encoder: JSONEncoder = JSONEncoder()) throws -> [String: Any] {
        let data = try encoder.encode(self)
        let object = try JSONSerialization.jsonObject(with: data)
        if let json = object as? [String: Any]  { return json }
        
        let context = DecodingError.Context(codingPath: [], debugDescription: "Deserialized object is not a dictionary")
        throw DecodingError.typeMismatch(type(of: object), context)
    }
}
